Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTree.java

```diff
@@ -0,0 +1,538 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.tree;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkNotNull;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.util.Arrays;
+import java.util.Random;
+
+import com.amazon.randomcutforest.CommonUtils;
+import com.amazon.randomcutforest.MultiVisitor;
+import com.amazon.randomcutforest.Sequential;
+import com.amazon.randomcutforest.Visitor;
+import com.amazon.randomcutforest.store.IPointStore;
+import com.amazon.randomcutforest.store.LeafStore;
+import com.amazon.randomcutforest.store.NodeStore;
+
+/**
+ * A Compact Random Cut Tree is a tree data structure whose leaves represent
+ * points inserted into the tree and whose interior nodes represent regions of
+ * space defined by Bounding Boxes and Cuts. New nodes and leaves are added to
+ * the tree by making random cuts.
+ *
+ * The offsets are encoded as follows: an offset greater or equal maxSize
+ * corresponds to a leaf node of offset (offset - maxSize) otherwise the offset
+ * corresponds to an internal node
+ *
+ * The main use of this class is to be updated with points sampled from a
+ * stream, and to define traversal methods. Users can then implement a
+ * {@link Visitor} which can be submitted to a traversal method in order to
+ * compute a statistic from the tree.
+ */
+public class CompactRandomCutTree implements ITree<Sequential<Integer>> {
```

Comment:
remark. halfway through tree implementation. some general issues other than comments. there seems to be some data type inconsistencies such as float[]/double[], int/short. is the plan to later iron them out eliminating data type casting?

Reply:
Yes, exactly. When I created this PR we still hadn't settled the 1.0/2.0 question, and I didn't want to change existing classes like Node and BoundingBox. Given the size of this PR, I'd like to leave the inconsistencies for now and fix them in a later PR (possibly the next one).